File: ids/src/protocols/l3/ipv4.cc
#include "ipv4.h"
#include "event_mgr.h"

namespace netos {

namespace ids {

uint16_t ipv4_hdr::checksum(std::shared_ptr<packet_buf> &pkt_buf)
{
    uint32_t chksum32 = 0;
    uint16_t chksum = 0;
    uint32_t i = 0;

    for (i = this->start_off; i < this->end_off; i += 2) {
        chksum32 += ((pkt_buf->buf_[i + 1] << 8) | (pkt_buf->buf_[i]));
    }

    chksum = ((chksum32 >> 16) | (chksum32));

    return ~chksum;
}

netos_status ipv4_hdr::serialize(std::shared_ptr<packet_buf> &pkt_buf)
{
    this->start_off = pkt_buf->offset_;

    pkt_buf->buf_[pkt_buf->offset_] = (this->version << 4);
    pkt_buf->buf_[pkt_buf->offset_] |= this->ihl;
    pkt_buf->offset_ ++;

    pkt_buf->buf_[pkt_buf->offset_] = (this->dscp << 2);
    pkt_buf->buf_[pkt_buf->offset_] |= this->ecn;
    pkt_buf->offset_ ++;

    pkt_buf->serialize_2_bytes(this->total_len);
    pkt_buf->serialize_2_bytes(this->id);
    pkt_buf->buf_[pkt_buf->offset_] = 0;
    if (this->flags.reserved) {
        pkt_buf->buf_[pkt_buf->offset_] |= 0x80;
    }
    if (this->flags.df) {
        pkt_buf->buf_[pkt_buf->offset_] |= 0x40;
    }
    if (this->flags.mf) {
        pkt_buf->buf_[pkt_buf->offset_] |= 0x20;
    }
    pkt_buf->buf_[pkt_buf->offset_] |= (this->frag_off & 0x0F00) >> 8;
    pkt_buf->buf_[pkt_buf->offset_ + 1] = this->frag_off & 0x00FF;
    pkt_buf->offset_ += 2;

    pkt_buf->serialize_byte(this->ttl);
    pkt_buf->serialize_byte(this->protocol);

    this->checksum_off = pkt_buf->offset_;

    pkt_buf->serialize_2_bytes(this->hdr_chksum);
    pkt_buf->serialize_4_bytes(this->src_addr);
    pkt_buf->serialize_4_bytes(this->dst_addr);

    this->end_off = pkt_buf->offset_;

    return netos_status::NETOS_STATUS_SUCCESS;
}

netos_status ipv4_hdr::deserialize(std::shared_ptr<packet_buf> &pkt_buf)
{
    this->version = (pkt_buf->buf_[pkt_buf->offset_] & 0xF0) >> 4;
    if (this->version != NETOS_IPV4_VERSION) {
        event_mgr::instance()->insert_event(IDS_EVENT_TYPE_DENY,
                                            event_description::EVENT_DESC_INVAL_IPV4_VERSION,
                                            event_protocol_level::EVENT_PROTOCOL_L3_IPV4,
                                            pkt_buf->len_);
        return netos_status::NETOS_STATUS_MALFORMED_PKT;
    }

    this->ihl = (pkt_buf->buf_[pkt_buf->offset_] & 0x0F);
    pkt_buf->offset_ ++;

    this->dscp = (pkt_buf->buf_[pkt_buf->offset_] & 0xFC) >> 2;
    this->ecn = (pkt_buf->buf_[pkt_buf->offset_] & 0x03);
    pkt_buf->offset_ ++;

    pkt_buf->deserialize_2_bytes(&this->total_len);
    pkt_buf->deserialize_2_bytes(&this->id);
    this->flags.reserved = !!(pkt_buf->buf_[pkt_buf->offset_] & 0x80);
    this->flags.df = !!(pkt_buf->buf_[pkt_buf->offset_] & 0x40);
    this->flags.mf = !!(pkt_buf->buf_[pkt_buf->offset_] & 0x20);
    pkt_buf->offset_ ++;

    this->frag_off = ((pkt_buf->buf_[pkt_buf->offset_] & 0x0F) << 4);
    this->frag_off |= pkt_buf->buf_[pkt_buf->offset_];
    pkt_buf->offset_ += 2;

    pkt_buf->deserialize_byte(&this->ttl);
    pkt_buf->deserialize_byte(&this->protocol);
    pkt_buf->deserialize_2_bytes(&this->hdr_chksum);
    pkt_buf->deserialize_4_bytes(&this->src_addr);
    pkt_buf->deserialize_4_bytes(&this->dst_addr);

    if (this->checksum(pkt_buf) != 0) {
        return netos_status::NETOS_STATUS_MALFORMED_PKT;
    }

    return netos_status::NETOS_STATUS_SUCCESS;
}

}

}

File: rust/protocols/src/ipv4.rs
#![allow(non_camel_case_types)]

use std::fmt;

use crate::pkt_buffer;

pub struct ipv4_hdr {
    pub version         : u8, // 4 bits
    pub header_len      : u8, // 4 bits
    pub dscp            : u8, // 6 bits
    pub ecn             : u8, // 2 bits
    pub total_len       : u16,
    pub id              : u16,
    pub reserved        : u8,
    pub dont_frag       : u8,
    pub more_frag       : u8,
    pub frag_off        : u32,
    pub ttl             : u8,
    pub protocol        : u8,
    pub hdr_chksum      : u16,
    pub src_ipaddr      : u32,
    pub dst_ipaddr      : u32
}

impl std::fmt::Display for ipv4_hdr {
    fn fmt(&self, f : &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "IPv4: ")?;
        writeln!(f, "\t version: {}", self.version)?;
        writeln!(f, "\t header_len: {}", self.header_len)?;
        writeln!(f, "\t dscp: {}", self.dscp)?;
        writeln!(f, "\t ecn: {}", self.ecn)?;
        writeln!(f, "\t total_len: {}", self.total_len)?;
        writeln!(f, "\t id: {}", self.id)?;
        writeln!(f, "\t reserved: {}", self.reserved)?;
        writeln!(f, "\t dont_frag: {}", self.dont_frag)?;
        writeln!(f, "\t more_frag: {}", self.more_frag)?;
        writeln!(f, "\t frag_off: {}", self.frag_off)?;
        writeln!(f, "\t ttl: {}", self.ttl)?;
        writeln!(f, "\t protocol: {}", self.protocol)?;
        writeln!(f, "\t hdr_chksum: {}", self.hdr_chksum)?;
        writeln!(f, "\t src_ipaddr: {:08x}", self.src_ipaddr)?;
        write!(f, "\t dst_ipaddr: {:08x}", self.dst_ipaddr)
    }
}

impl ipv4_hdr {
    pub fn new() -> Self {
        Self {
            version     : 0,
            header_len  : 0,
            dscp        : 0,
            ecn         : 0,
            total_len   : 0,
            id          : 0,
            reserved    : 0,
            dont_frag   : 0,
            more_frag   : 0,
            frag_off    : 0,
            ttl         : 0,
            protocol    : 0,
            hdr_chksum  : 0,
            src_ipaddr  : 0,
            dst_ipaddr  : 0
        }
    }

    pub fn decode(&mut self, pkt_buf : &mut pkt_buffer::netos_pkt_buffer) -> i32 {
        self.version = (pkt_buf.buffer[pkt_buf.offset] & 0xF0) >> 4;
        self.header_len = (pkt_buf.buffer[pkt_buf.offset] & 0x0F) * 5;
        pkt_buf.offset += 1;

        self.dscp = (pkt_buf.buffer[pkt_buf.offset] & 0xFC) >> 2;
        self.ecn = pkt_buf.buffer[pkt_buf.offset] & 0x03;

        0
    }
}

